File: SpaceDock/endpoints/publisher.py
from datetime import datetime
from flask import request
from sqlalchemy import desc
from SpaceDock.common import edit_object, user_has, with_session
from SpaceDock.database import db
from SpaceDock.formatting import publisher_info
from SpaceDock.objects import Publisher
from SpaceDock.routing import route


@route('/api/publishers')
def publishers_list():
    """
    Outputs all publishers known by the application
    """
    results = dict()
    for v in Publisher.query.order_by(desc(Publisher.id)).all():
        results[v.id] = v.name
    return {"error": False, 'count': len(results), 'data': results}

@route('/api/publishers/<pubid>')
def publishers_info(pubid):
    """
    Outputs detailed infos for one publisher
    """
    if not pubid.isdigit() or not Publisher.query.filter(Publisher.id == int(pubid)).first():
        return {'error': True, 'reasons': ['Invalid publisher ID'], 'codes': ['2110']}, 400
    # Return the info
    pub = Publisher.query.filter(Publisher.id == int(pubid)).first()
    return {'error': False, 'count': 1, 'data': publisher_info(pub)}

@route('/api/publishers/<pubid>/edit', methods=['POST'])
@user_has('publisher-edit', params=['pubid'])
@with_session
def edit_publisher(pubid):
    """
    Edits a publisher, based on the request parameters. Required fields: data
    """
    if not pubid.isdigit() or not Publisher.query.filter(Publisher.id == int(pubid)).first():
        return {'error': True, 'reasons': ['Invalid publisher ID'], 'codes': ['2110']}, 400

    # Get the matching game and edit it
    pub = Publisher.query.filter(Publisher.id == int(pubid)).first()
    code = edit_object(pub, request.json)
    
    # Error check
    if code == 3:
        return {'error': True, 'reasons': ['The value you submitted is invalid'], 'codes': ['2180']}, 400
    elif code == 2:
        return {'error': True, 'reasons': ['You tried to edit a value that doesn\'t exist.'], 'codes': ['3090']}, 400
    elif code == 1:
        return {'error': True, 'reasons': ['You tried to edit a value that is marked as read-only.'], 'codes': ['3095']}, 400
    else:
        pub.updated = datetime.now()
        return {'error': False, 'count': 1, 'data': publisher_info(pub)}

@route('/api/publishers/add', methods=['POST'])
@user_has('publisher-add')
@with_session
def add_publisher():
    """
    Adds a publisher, based on the request parameters. Required fields: name
    """
    # Get variables
    name = request.json.get('name')

    # Check for existence
    if Publisher.query.filter(Publisher.name == name).first():
        return {'error': True, 'reasons': ['A publisher with this name already exists.'], 'codes': ['2000']}, 400

    # Get the matching game and edit it
    pub = Publisher(name)
    db.add(pub)
    db.flush()
    return {'error': False, 'count': 1, 'data': publisher_info(pub)}

@route('/api/publishers/remove', methods=['POST'])
@user_has('publisher-remove')
@with_session
def remove_publisher():
    """
    Removes a game from existence. Required fields: pubid
    """
    pubid = request.json.get('pubid')

    # Check if the pubid is valid
    if not isinstance(pubid, int) or not Publisher.query.filter(Publisher.id == pubid).first():
        return {'error': True, 'reasons': ['Invalid publisher ID'], 'codes': ['2110']}, 400

    # Get the publisher and remove it
    pub = Publisher.query.filter(Publisher.id == pubid).first()
    db.delete(pub)
    return {'error': False}

File: src/SpaceDock/routes/publisher.go
/*
 SpaceDock Backend
 API Backend for the SpaceDock infrastructure to host modfiles for various games

 SpaceDock Backend is licensed under the Terms of the MIT License.
 Copyright (c) 2017 Dorian Stoll (StollD), RockyTV
*/

package routes

import (
    "SpaceDock"
    "SpaceDock/middleware"
    "SpaceDock/objects"
    "SpaceDock/utils"
    "github.com/spf13/cast"
    "gopkg.in/kataras/iris.v6"
)

/*
 Registers the routes for the publisher section
 */
func PublisherRegister() {
    Register(GET, "/api/publishers", publishers_list)
    Register(GET, "/api/publishers/:pubid", publishers_info)
    Register(PUT, "/api/publishers/:pubid",
        middleware.NeedsPermission("publisher-edit", true, "pubid"),
        edit_publisher,
    )
    Register(POST, "/api/publishers",
        middleware.NeedsPermission("publisher-add", true),
        add_publisher,
    )
    Register(DELETE, "/api/publishers/:pubid",
        middleware.NeedsPermission("publisher-remove", true),
        remove_publisher,
    )
}

/*
 Path: /api/publishers
 Method: GET
 Description: Outputs all publishers known by the application
 */
func publishers_list(ctx *iris.Context) {
    var publishers []objects.Publisher
    SpaceDock.Database.Find(&publishers)
    output := make([]map[string]interface{}, len(publishers))
    for i,element := range publishers {
        output[i] = utils.ToMap(element)
    }
    utils.WriteJSON(ctx, iris.StatusOK, iris.Map{"error": false, "count": len(publishers), "data": output})
}

/*
 Path: /api/publishers/:pubid
 Method: GET
 Description: Outputs detailed infos for one publisher
 */
func publishers_info(ctx *iris.Context) {
    pubid := cast.ToUint(ctx.GetString("pubid"))

    // Get the publisher
    pub := &objects.Publisher{}
    SpaceDock.Database.Where("id = ?", pubid).First(pub)
    if pub.ID != pubid {
        utils.WriteJSON(ctx, iris.StatusNotFound, utils.Error("The publisher ID is invalid").Code(2110))
        return
    }

    // Return the info
    utils.WriteJSON(ctx, iris.StatusOK, iris.Map{"error": false, "count": 1, "data": utils.ToMap(pub)})
}

/*
 Path: /api/publishers/:pubid
 Method: PUT
 Description: Edits a publisher, based on the request parameters. Required fields: data
 Abilities: publisher-edit
 */
func edit_publisher(ctx *iris.Context) {
    pubid := cast.ToUint(ctx.GetString("pubid"))

    // Get the publisher
    pub := &objects.Publisher{}
    SpaceDock.Database.Where("id = ?", pubid).First(pub)
    if pub.ID != pubid {
        utils.WriteJSON(ctx, iris.StatusNotFound, utils.Error("The publisher ID is invalid").Code(2110))
        return
    }

    // Edit the publisher
    code := utils.EditObject(pub, utils.GetFullJSON(ctx))
    if code == 3 {
        utils.WriteJSON(ctx, iris.StatusBadRequest, utils.Error("The value you submitted is invalid").Code(2180))
        return
    } else if code == 2 {
        utils.WriteJSON(ctx, iris.StatusBadRequest, utils.Error("You tried to edit a value that doesn't exist.").Code(3090))
        return
    } else if code == 1 {
        utils.WriteJSON(ctx, iris.StatusBadRequest, utils.Error("You tried to edit a value that is marked as read-only.").Code(3095))
        return
    }
    SpaceDock.Database.Save(pub)
    utils.WriteJSON(ctx, iris.StatusOK, iris.Map{"error": false, "count": 1, "data": utils.ToMap(pub)})
}

/*
 Path: /api/publishers/
 Method: POST
 Description: Adds a publisher, based on the request parameters. Required fields: name
 Abilities: publisher-add
 */
func add_publisher(ctx *iris.Context) {
    name := cast.ToString(utils.GetJSON(ctx, "name"))

    // Get the publisher
    pub := &objects.Publisher{}
    SpaceDock.Database.Where("name = ?", name).First(pub)
    if pub.Name == name {
        utils.WriteJSON(ctx, iris.StatusBadRequest, utils.Error("A publisher with this name already exists.").Code(2000))
        return
    }

    // Add the publisher
    pub = objects.NewPublisher(name)
    SpaceDock.Database.Save(pub)
    utils.WriteJSON(ctx, iris.StatusOK, iris.Map{"error": false, "count": 1, "data": utils.ToMap(pub)})
}

/*
 Path: /api/publishers/:pubid
 Method: DELETE
 Description: Removes a game from existence. Required fields: pubid
 Abilities: publisher-remove
 */
func remove_publisher(ctx *iris.Context) {
    pubid := cast.ToUint(ctx.GetString("pubid"))

    // Get the publisher
    pub := &objects.Publisher{}
    SpaceDock.Database.Where("id = ?", pubid).First(pub)
    if pub.ID != pubid {
        utils.WriteJSON(ctx, iris.StatusNotFound, utils.Error("The publisher ID is invalid").Code(2110))
        return
    }

    // Delete the publisher
    SpaceDock.Database.Delete(pub)
    utils.WriteJSON(ctx, iris.StatusOK, iris.Map{"error": false})
}